First commits

jsf9k opened the pull request:
## 🗣 Description ##

This pull request contains the initial commits for this project.

## 💭 Motivation and context ##

The functionality of [cisagov/guacscanner](https://github.com/cisagov/guacscanner) must be integrated into [cisagov/guacamole-composition](https://github.com/cisagov/guacamole-composition) for the RPT UAT taking place in the COOL.

See also:
- cisagov/guacscanner/pull/1
- cisagov/guacamole-composition#30
- cisagov/ansible-role-guacamole#38
- cisagov/guacamole-packer#53
- cisagov/cool-assessment-terraform#149

## 🧪 Testing ##

I have applied these changes to env0 in our staging COOL environment and verified that they function as expected.

## ✅ Checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] _All_ future TODOs are captured in issues, which are referenced in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.
- [x] Finalize version.

jsf9k committed:
Bump version from 0.0.1 to 0.0.1-rc.1

jsf9k committed:
Update secrets

jsf9k committed:
Update Dockerfile and Docker composition

jsf9k committed:
Update tests

jsf9k committed:
Add more secrets

jsf9k committed:
Correct image name

jsf9k committed:
Fix typo

jsf9k committed:
Tweak test to make it pass

jsf9k committed:
Bump version from 0.0.1-rc.1 to 0.0.1-rc.2

jsf9k committed:
Bump version from 0.0.1-rc.2 to 0.0.1-rc.3

jsf9k committed:
Bump version from 0.0.1-rc.3 to 0.0.1-rc.4

jsf9k committed:
Bump version from 0.0.1-rc.4 to 0.0.1-rc.5

jsf9k committed:
Bump version from 0.0.1-rc.5 to 0.0.1-rc.6

jsf9k committed:
Bump version from 0.0.1-rc.6 to 0.0.1-rc.7

jsf9k committed:
Temporarily install some packages to aid in debugging

Also give CISA_USER sudo rights (again temporarily).

jsf9k committed:
Bump version from 0.0.1-rc.7 to 0.0.1-rc.8

jsf9k committed:
Bump version from 0.0.1-rc.8 to 0.0.1-rc.9

jsf9k committed:
Bump version from 0.0.1-rc.9 to 0.0.1-rc.10

jsf9k committed:
Bump version from 0.0.1-rc.10 to 0.0.1-rc.11

jsf9k committed:
Bump version from 0.0.1-rc.11 to 0.0.1-rc.12

jsf9k committed:
Bump version from 0.0.1-rc.12 to 0.0.1-rc.13

jsf9k committed:
Update README from skeleton content

jsf9k committed:
Bump version from 0.0.1-rc.13 to 0.0.1-rc.14

jsf9k committed:
Bump version from 0.0.1-rc.14 to 0.0.1-rc.15

jsf9k committed:
Bump version from 0.0.1-rc.15 to 0.0.1-rc.16

jsf9k committed:
Bump version from 0.0.1-rc.16 to 0.0.1-rc.17

jsf9k committed:
Bump version from 0.0.1-rc.17 to 0.0.1-rc.18

dav3r commented on README.md:
```diff
@@ -125,46 +82,49 @@ environment variables.  See the
-|`cisagov/example:edge` | The most recent image built from a merge into the `develop` branch of this repository. |
-|`cisagov/example:nightly` | A nightly build of the `develop` branch of this repository. |
-|`cisagov/example:latest`| The most recent release image pushed to a container registry.  Pulling an image using the `:latest` tag [should be avoided.](https://vsupalov.com/docker-latest-tag/) |
+|`cisagov/guacscanner:1.2.3`| An exact release version. |
+|`cisagov/guacscanner:1.2`| The most recent release matching the major and minor version numbers. |
+|`cisagov/guacscanner:1`| The most recent release matching the major version number. |
```
Match current version:
```suggestion
|`cisagov/guacscanner:0.0.1`| An exact release version. |
|`cisagov/guacscanner:0.0`| The most recent release matching the major and minor version numbers. |
|`cisagov/guacscanner:0`| The most recent release matching the major version number. |
```

jsf9k replied:
I know we do this elsewhere, but looking at it this time I think the default values (because the major, minor, and patch values are all different from one another) make the meaning behind each of the the tags clearer.  If the version is `0.0.1`, and you check out `cisagov/guacscanner:0`, it isn't clear a priori that the zero in the image tag refers to the major version.

This is a subtle point, but I think it is an important one.

dav3r replied:
I see what you're saying, but I still think that this is clear when coupled with the corresponding explanatory description text.  I like always showing in the docs what the current versions are, rather than tags that may or may not exist or be the latest versions.  I think we need @mcdonnnj and @felddy to weigh in here.

jsf9k replied:
I'm not convinced it's best, but I will defer to the inertia of the way we have been doing it.   Please enjoy commit 2446f05.

dav3r replied:
Maybe open an issue or discussion for this in skeleton-docker?

dav3r commented on README.md:
```diff
@@ -180,25 +140,33 @@ There are no required environment variables.
-| `quote.txt` | Replaces the secret stored in the example library's package data. |
+| postgres-username | Text file containing the username of the `postgres` user used by the `guacamole` container |
+| postgres-password | Text file containing the password of the `postgres` user used by the `guacamole` container |
+| private_ssh_key | Text file containing the private ssh key to use for SFTP file transfer in Guacamole. |
```
Capitalization:
```suggestion
| private_ssh_key | Text file containing the private SSH key to use for SFTP file transfer in Guacamole. |
```

dav3r commented on tests/container_test.py:
```diff
@@ -26,33 +19,36 @@ def test_container_count(dockerc):
+#         raise Exception(
+#             f"Container does not seem ready.  "
+#             f'Expected "{READY_MESSAGE}" in the log within {TIMEOUT} seconds.'
+#         )
```
Are you planning to implement something similar to these tests?  If so, there should be a TODO and an issue.

jsf9k replied:
Please see commit f6f77d0.

jsf9k committed:
Change guacscanner restart policy to always

This will cause guacscanner to be restarted if it throws an unhandled
exception; otherwise, the container just ceases to run and stops
processing Guacamole connections as instances appear and disappear.

jsf9k committed:
Make capitalization of "SSH" consistent

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Add TODOs to implement a few standard tests that are currently commented out

jsf9k committed:
Bump version from 0.0.1-rc.18 to 0.0.1

jsf9k committed:
Make example version tags agree with latest version in repo

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

dav3r reviewed (approved):
🐧

jsf9k committed:
Remove default log level setting

It logs too much stuff and eats up disk space on the host.

felddy commented on Dockerfile:
```diff
@@ -1,38 +1,86 @@
+# We need redis-tools so we can use redis-cli to communicate with
+# redis.  wget is used inside of gather-domains.sh.
+#
+# Install dependencies are only needed for software installation and
```
This should be implemented as multi-stage build instead of trying to clean up. 
The first stage should do all the installation and the second stage should copy what it needs from the first. 

See: https://docs.docker.com/develop/develop-images/multistage-build/

I can help with this if you'd like.

felddy replied:
I have a current project (not ⚔️ 🐉) where I did something very similar.  I think it would be a useful example: 

https://github.com/felddy/weewx-docker/blob/develop/Dockerfile

jsf9k replied:
I had the same thought when creating the `Dockerfile`, but at the time I was trying to get something going that I could test.  I shall make this change.

jsf9k replied:
@cisagov/team-ois - please have a look at commit cf86400 and let me know what you think.

felddy replied:
Wunderbar!

felddy reviewed (changes requested):
Looking good.  Please see my comment about multi-stage builds.

jsf9k committed:
Remove temporary code that is no longer of use

jsf9k committed:
Make Dockerfile a multistage build

@felddy says this is preferable to using a single-stage build and
uninstalling the install dependencies.  I agree with him in the
general case, but it definitely complicated things here.

dav3r commented on Dockerfile:
```diff
@@ -1,38 +1,117 @@
 # https://github.com/opencontainers/image-spec/blob/master/annotations.md
 # Note: Additional labels are added by the build workflow.
-LABEL org.opencontainers.image.authors="mark.feldhousen@cisa.dhs.gov"
+LABEL org.opencontainers.image.authors="jeremy.frasier@cisa.dhs.gov"
```
Do we care about the `LABEL`s anywhere except in the final image?

jsf9k replied:
I don't know...I figured it wouldn't hurt anything to have them there.  I'm happy to remove them if @cisagov/team-ois thinks they should not be there.

felddy replied:
I'd say ditch em.  Can't have a bug in a line of code that doesn't exist.  
Plus I could totally see me updating this in the future and not understanding why the final image tags don't update.

jsf9k replied:
Enjoy commit 8f2554a!

felddy replied:
So hot!  🌶️

dav3r reviewed (approved):
Nice multi-stage action!  👍  I left one question, but nothing that is in the way of approval.

mcdonnnj commented on Dockerfile:
```diff
@@ -1,38 +1,117 @@
+RUN python -m pip install --upgrade pip setuptools wheel
+
+# Download and install guacscanner
+RUN wget -O sourcecode.tgz \
+    https://github.com/cisagov/guacscanner/archive/v${VERSION}.tar.gz && \
+    tar xzf sourcecode.tgz --strip-components=1 && \
+    python -m pip install --requirement requirements.txt && \
+    rm sourcecode.tgz
```
```suggestion
RUN pip install --no-cache-dir https://github.com/cisagov/guacscanner/archive/v${VERSION}.tar.gz
```

dav3r replied:
This change should really make it's way into `skeleton-docker`.

mcdonnnj replied:
I agree. Updating some of our skeleton stuff to be both more modern and align with how we currently develop takes up a sizeable amount of my backlog.

mcdonnnj commented on Dockerfile:
```diff
@@ -1,38 +1,117 @@
+ENV DEPS \
+    libpq-dev
+ENV INSTALL_DEPS \
+    wget
```
These should be pinned.

felddy replied:
@mcdonnnj is running a tight ship.  

I'd be fine with these being unpinned since they are tied to the base image.   Not saying don't do it, but I'd probably let these float.   

You know how I am.  ☮️

mcdonnnj replied:
They are only tied to what `apt update` pulls down when the image is built. A reproducible Docker build needs to have it all pinned x)

jsf9k replied:
Please see commit 0bf0e0f.

jsf9k committed:
Bump version from 0.0.1 to 0.0.2

jsf9k committed:
Bump version from 0.0.2 to 0.0.2-rc.1

mcdonnnj commented on Dockerfile:
```diff
@@ -1,38 +1,117 @@
+# Dependencies
+###
+ENV DEPS \
+    libpq-dev
```
This should be pinned.

jsf9k replied:
Please see commit 0bf0e0f.

mcdonnnj commented on Dockerfile:
```diff
@@ -1,38 +1,117 @@
+RUN python -m venv ${PY_VENV}
+ENV PATH="${PY_VENV}/bin:$PATH"
+# Install/upgrade core Python dependencies
+RUN python -m pip install --upgrade pip setuptools wheel
```
We should pin these packages.

mcdonnnj replied:
```suggestion
RUN python -m pip install --no-cache-dir --upgrade pip setuptools wheel
```

jsf9k replied:
Please also see commit 33a282c.

mcdonnnj commented on README.md:
```diff
@@ -125,46 +82,49 @@ environment variables.  See the
+
+<!-- | Mount point | Purpose        | -->
+<!-- |-------------|----------------| -->
+<!-- | `/var/log`  |  Log storage   | -->
```
I think it's better to use a single logical comment block. This also mirrors how the required environment variable section is commented out.

```suggestion
<!--
| Mount point | Purpose        |
|-------------|----------------|
| `/var/log`  |  Log storage   |
-->
```

mcdonnnj commented on README.md:
```diff
@@ -125,46 +82,49 @@ environment variables.  See the
-| Port | Purpose        |
-|------|----------------|
-| 8080 | Example only; nothing is actually listening on the port |
+<!-- | Port | Purpose        | -->
+<!-- |------|----------------| -->
+<!-- | 8080 | Example only; nothing is actually listening on the port | -->
```
```suggestion
<!--
| Port | Purpose        |
|------|----------------|
| 8080 | Example only; nothing is actually listening on the port |
-->
```

mcdonnnj commented on README.md:
```diff
@@ -125,46 +82,49 @@ environment variables.  See the
 
-The sample [Docker composition](docker-compose.yml) publishes the
-exposed port at 8080.
+<!-- The sample [Docker composition](docker-compose.yml) publishes the -->
+<!-- exposed port at 8080. -->
```
```suggestion
<!--
The sample [Docker composition](docker-compose.yml) publishes the
exposed port at 8080.
-->
```

mcdonnnj commented on README.md:
```diff
@@ -180,25 +140,33 @@ There are no required environment variables.
-| `ECHO_MESSAGE` | Sets the message echoed by this container.  | `Hello World from Dockerfile` |
+There are no required environment variables.
+
+<!-- | Name  | Purpose | Default | -->
+<!-- |-------|---------|---------| -->
+<!-- | `ECHO_MESSAGE` | Sets the message echoed by this container.  | `Hello World from Dockerfile` | -->
```
```suggestion
<!--
| Name  | Purpose | Default |
|-------|---------|---------|
| `ECHO_MESSAGE` | Sets the message echoed by this container.  | `Hello World from Dockerfile` |
-->
```

mcdonnnj reviewed (changes requested):
Strong work 💪💪💪 I did have some feedback for your consideration.

jsf9k committed:
Remove labels in compile stage

They would never see the light of day anyway.

felddy commented on Dockerfile:
```diff
@@ -1,38 +1,117 @@
+
+# Clean up aptitude cruft
+RUN apt-get clean
+RUN rm -rf /var/lib/apt/lists/*
```
```suggestion
RUN apt-get update && \
    apt-get install --no-install-recommends --no-install-suggests --yes \
    ${DEPS} && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*
```

Consider unifying these three `RUN` instructions.  This will reduce the number of layers in the image, and reduce its final size.  As it stands, the cruft will still be in there, but just shadowed by the UnionFS. 

https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers

jsf9k replied:
Please enjoy commit 82edf1c!

felddy replied:
Stellar! 🌟

jsf9k committed:
Install guacscanner via a simpler command

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Pin Debian packages

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

felddy reviewed (approved):
This PR represents the best of what we have to offer! 🚀 🌕
Let me get out of your way! 

>“Apollo succeeded at critical moments like this because the bosses had no hesitation about assigning crucial tasks to one individual, trusting his judgment, and then getting out of his way.” --Gene Kranz

jsf9k committed:
Add --no-cache-dir to pip install command

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Pin code Python dependencies installed via pip

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

dav3r commented:
> > “Apollo succeeded at critical moments like this because the bosses had no hesitation about assigning crucial tasks to one individual, trusting his judgment, and then getting out of his way.” --Gene Kranz

https://www.youtube.com/watch?v=sngs6wj8tEY&t=26s

![image](https://user-images.githubusercontent.com/21343441/141189445-4b1a702f-f81f-4bbc-932f-1c89c5b5e963.png)

jsf9k committed:
Use a different style of HTML comment fencing

The new style agrees better with that is already in the skeleton.

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Combine aptitude cleanup with aptitude install step

@felddy pointed out that this must be done in one fell swoop to
actually reduce the size of the resulting Docker image:
https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers

Co-authored-by: Mark Feldhousen <felddy@users.noreply.github.com>

jsf9k commented:
> This PR represents the best of what we have to offer! rocket full_moon Let me get out of your way!

I think you left out a word at the beginning:  "_Sadly,_  this PR represents the best of what we have to offer..."  :laughing:

jsf9k commented:
Pinning the Debian package `wget` is tricky because [the package has a different version depending on the architecture](https://packages.debian.org/bullseye/wget).  Does anyone know how to handle this case cleanly in the `Dockerfile`?

felddy commented:
> Pinning the Debian package `wget` is tricky because [the package has a different version depending on the architecture](https://packages.debian.org/bullseye/wget). Does anyone know how to handle this case cleanly in the `Dockerfile`?

I think a comment to that effect would suffice.  ;)

jsf9k commented:
> > Pinning the Debian package `wget` is tricky because [the package has a different version depending on the architecture](https://packages.debian.org/bullseye/wget). Does anyone know how to handle this case cleanly in the `Dockerfile`?
> 
> I think a comment to that effect would suffice. ;)

I think it's more complicated than that, since we build Docker images for lots of different architectures.  This is precisely why the GH Actions are failing.  In this case, I need to use version `1.21-1+b1` for `amd64` and version `1.21-1` otherwise.  I presume the solution is to make use of [this jazz](https://docs.docker.com/engine/reference/builder/#automatic-platform-args-in-the-global-scope) somehow, but I don't see a way to do ternary logic with `ENV`s in a `Dockerfile`.

[Here](https://stackoverflow.com/questions/67596193/building-a-multi-architecture-docker-image-but-dockerfile-requires-different-pa) is a post from StackOverflow where someone asks a similar question.

jsf9k commented:
> > Pinning the Debian package `wget` is tricky because [the package has a different version depending on the architecture](https://packages.debian.org/bullseye/wget). Does anyone know how to handle this case cleanly in the `Dockerfile`?
> 
> I think a comment to that effect would suffice. ;)

Ah, I see what you mean.  I will comment and avoid pinning for now.

jsf9k committed:
Avoid pinning the package version for wget

I'd like to pin the version of wget to keep the build reproducible,
but it's tricky.

I need to use version 1.21-1+b1 of wget for amd64 and version 1.21-1
of wget otherwise.
https://packages.debian.org/bullseye/wget

I presume the solution is to somehow make use of this jazz:
https://docs.docker.com/engine/reference/builder/#automatic-platform-args-in-the-global-scope)

But I don't see a way to do ternary logic with ENVs in a Dockerfile.

Here is a post from StackOverflow where someone asks a similar
question: https://stackoverflow.com/questions/67596193/building-a-multi-architecture-docker-image-but-dockerfile-requires-different-pa

jsf9k commented:
> This PR represents the best of what we have to offer! rocket full_moon Let me get out of your way!
> 
> > “Apollo succeeded at critical moments like this because the bosses had no hesitation about assigning crucial tasks to one individual, trusting his judgment, and then getting out of his way.” --Gene Kranz

Kranz had a lot of great quotes and in my opinion is a good example of a great leader in times of crisis.  He gave [a short but great speech, called the "Kranz Dictum"](https://en.wikipedia.org/wiki/Gene_Kranz#%22The_Kranz_Dictum%22), the morning after three astronauts died in a fire on the launch pad.

mcdonnnj commented on README.md:
```diff
@@ -180,25 +146,35 @@ There are no required environment variables.
 
 ### Optional ###
 
+There are no required environment variables.
```
```suggestion
There are no optional environment variables.
```

mcdonnnj commented on README.md:
```diff
@@ -180,25 +146,35 @@ There are no required environment variables.
 | Filename     | Purpose |
 |--------------|---------|
-| `quote.txt` | Replaces the secret stored in the example library's package data. |
+| postgres-username | Text file containing the username of the `postgres` user used by the `guacamole` container |
+| postgres-password | Text file containing the password of the `postgres` user used by the `guacamole` container |
```
Just for consistency.

```suggestion
| postgres-username | Text file containing the username of the `postgres` user used by the `guacamole` container. |
| postgres-password | Text file containing the password of the `postgres` user used by the `guacamole` container. |
```

mcdonnnj reviewed (approved):
Thanks again for your work on this enhancement. I just have a couple suggestions related to the README.

jsf9k committed:
Fix typo

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Add periods for consistency

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Bump version from 0.0.2-rc.1 to 0.0.2-rc.2

jsf9k committed:
Bump the version of libpq

The old version is resulting in a build error, while the new version
succeeds.

jsf9k committed:
Bump version from 0.0.2-rc.2 to 0.0.2

jsf9k committed:
Bump version from 0.0.2 to 1.0.0
